src/server/run-websocket-server.ts

```diff
@@ -1,30 +1,60 @@
 import * as WebSocket from "ws";
 import * as Knex from "knex";
+import { EventEmitter } from "events";
 import { JsonRpcRequest } from "../types";
 import { isJsonRpcRequest } from "./is-json-rpc-request";
 import { dispatchJsonRpcRequest } from "./dispatch-json-rpc-request";
 import { makeJsonRpcResponse } from "./make-json-rpc-response";
+import { makeJsonRpcError, JsonRpcErrorCode } from "./make-json-rpc-error";
+import { Subscriptions } from "./subscriptions";
 
 export function runWebsocketServer(db: Knex, port: number): void {
   console.log("Starting websocket server on port", port);
   const websocketServer: WebSocket.Server = new WebSocket.Server({ port });
   websocketServer.on("connection", (websocket: WebSocket): void => {
+    const subscriptions = new Subscriptions();
+
     websocket.on("message", (data: WebSocket.Data): void => {
+      let message: any;
+      try {
+        message = JSON.parse(data as string);
+        if (!isJsonRpcRequest(message)) return console.error("bad json rpc message received:", message);
+      } catch (exc) {
+        return websocket.send(makeJsonRpcError("-1", JsonRpcErrorCode.ParseError, "Bad JSON RPC Message Received", { originalText: data as string }));
+      }
+
       try {
-        const message = JSON.parse(data as string);
-        if (isJsonRpcRequest(message)) {
-          dispatchJsonRpcRequest(db, message as JsonRpcRequest, (err?: Error|null, result?: any): void => {
-            if (err) return console.error("dispatch error:", err);
+        if (message.method === "subscribe") {
+          const eventName: string = message.params.shift();
+
+          try {
+            const subscription: string = subscriptions.subscribe(eventName, message.params, (data: {}): void => {
+              websocket.send(makeJsonRpcResponse(message.id, { subscription, result: data }));
+            });
+            websocket.send(makeJsonRpcResponse(message.id, { subscription }));
+          } catch (exc) {
+            websocket.send(makeJsonRpcError(message.id, JsonRpcErrorCode.MethodNotFound, exc.toString(), false));
+          }
+        } else if (message.method === "unsubscribe") {
+          const subscription: string = message.params.shift();
+          subscriptions.unsubscribe(subscription);
+          websocket.send(makeJsonRpcResponse(message.id, true));
+        } else {
+          dispatchJsonRpcRequest(db, message as JsonRpcRequest, (err?: Error | null, result?: any): void => {
```

Comment:
In another PR, we changed most of the `err?` to `err`, couldn't find any instances where it would be `undefined`